### PR TITLE
Addition of function to calculate rally point instance in AWS Auto Scaling Groups

### LINKED ISSUE #32: Add a function for calculating a rally point instance as part of an AWS Auto Scaling Group
Similar to the logic [here](https://github.com/gruntwork-io/terraform-aws-couchbase/blob/master/modules/couchbase-commons/couchbase-rally-point), a function should exist that can be used to calculate a single "rally point" instance out of a pool of instances in an AWS Auto Scaling Group. This would be useful in cases where commands need to be run once as part of a bootstrapping process (e.g. something like hashicorp/terraform-aws-consul#119).

### PR DESCRIPTION
**Pull request by yardbirdsax**
Fixes #32 as well as updating the Dockerfiles used for testing to use Python 3 (I could not get them to run using the older Python version).

### CODE REVIEW

**Comment by yardbirdsax** [comment]
I believe I should split out the unit tests here for the private vs public DNS name usage, but wanted to get this up in the meantime for a review other than that.

**Comment by yardbirdsax** [comment]
I split the unit tests out and also refactored the new tests such that the setup code for each set isn't duplicated. The latter isn't the same way other things are done, so if you'd like me to back that out and just have the code in each test let me know. Thanks!

File: test/aws-wrapper.bats [hunk 1]
```diff
@@ -1311,4 +1311,233 @@ END_HEREDOC
...
+    }
+  ]
+}
+END_HEREDOC
+)
+
+  load_aws_mock "" "$asg" "$instances"
+}
+
+@test "aws_wrapper_get_asg_rally_point by launch time, private" {
```
Comment (brikis98) [inline test/aws-wrapper.bats:1517]: Nice tests! 👍 

Should there be tests that check how this function handles an empty ASG?

**Review by brikis98 — COMMENTED** [review]
Thanks for the PR!

File: Dockerfile.ubuntu16.04.bats [hunk 2]
```diff
@@ -12,13 +12,18 @@ RUN apt-get install -y software-properties-common && \
     apt-get install -y bats
 
 # Upgrade pip
-RUN pip install -U pip
+RUN sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 1 && \
+    sudo update-alternatives --config python && \
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply Dockerfile.ubuntu16.04.bats:16]: Likely not, I'll remove it and make sure things still work.

File: Dockerfile.ubuntu16.04.bats [hunk 3]
```diff
@@ -12,13 +12,18 @@ RUN apt-get install -y software-properties-common && \
...
+    curl https://bootstrap.pypa.io/pip/3.5/get-pip.py -o /tmp/get-pip.py && \
+    python /tmp/get-pip.py && \
+    pip install -U pip
 
 # Install AWS CLI
 RUN pip install awscli --upgrade --user
 
 # Install moto: https://github.com/spulec/moto
-RUN pip install flask moto moto[server] networkx==2.2
+RUN sudo apt install libffi-dev python3-dev -y && \
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply Dockerfile.ubuntu16.04.bats:25]: Done

File: Dockerfile.ubuntu16.04.bats [hunk 4]
```diff
@@ -12,13 +12,18 @@ RUN apt-get install -y software-properties-common && \
     apt-get install -y bats
 
 # Upgrade pip
-RUN pip install -U pip
+RUN sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 1 && \
+    sudo update-alternatives --config python && \
+    curl https://bootstrap.pypa.io/pip/3.5/get-pip.py -o /tmp/get-pip.py && \
+    python /tmp/get-pip.py && \
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply Dockerfile.ubuntu16.04.bats:18]: Done

File: Dockerfile.ubuntu18.04.bats [hunk 5]
```diff
@@ -12,11 +12,17 @@ RUN apt-get install -y software-properties-common && \
     apt-get install -y bats
 
 # Install AWS CLI
-RUN pip install awscli --upgrade --user
+RUN sudo apt install python3-distutils python3-apt -y && \
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply Dockerfile.ubuntu18.04.bats:15]: Done

File: modules/bash-commons/src/aws-wrapper.sh [hunk 6]
```diff
@@ -215,3 +215,35 @@ function aws_wrapper_get_hostname {
     aws_get_instance_private_hostname
   fi
 }
+
+# Calculates a rally point instance in an ASG, which is useful in cases where a piece of code needs to be run only
+# once upon provision of the ASG. Based off of https://github.com/gruntwork-io/terraform-aws-couchbase/blob/master/modules/couchbase-commons/couchbase-rally-point.
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply modules/bash-commons/src/aws-wrapper.sh:220]: Done

File: test/aws-wrapper.bats [hunk 7]
```diff
@@ -1311,4 +1311,233 @@ END_HEREDOC
...
   assert_success
   assert_equal "ip-10-251-50-12.ec2.internal" "$out"
+}
+
+function setup_rally_point_by_instance {
+  local readonly asg_name="$1"
+  local readonly aws_region="$2"
+  local readonly size=3
+
+  local readonly asg=$(cat <<END_HEREDOC
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply test/aws-wrapper.bats:1321]: Done, as well as other `local readonly` variables. Thanks for bearing with me as I learn the ins and outs of BATS. 😄

File: test/aws-wrapper.bats [hunk 8]
```diff
@@ -1311,4 +1311,233 @@ END_HEREDOC
...
+  local readonly aws_region="us-west-2"
+
+  setup_rally_point_by_instance $asg_name $aws_region
+
+  local out
+
+  out=$(aws_wrapper_get_asg_rally_point $asg_name $aws_region)
+  assert_success
+  assert_equal "$out" "ip-10-251-50-12.ec2.internal"
+
```
[the remark this answers is not in the document]
Comment (yardbirdsax) [reply test/aws-wrapper.bats:1413]: Fixed

File: test/aws-wrapper.bats (hunk 1 above)
Comment (yardbirdsax) [reply test/aws-wrapper.bats:1514]: Thanks! That's a good point, let me ponder on that and see what I can write. I've pushed up the changes for the rest of your comments.

Comment (yardbirdsax) [reply test/aws-wrapper.bats:1514]: I added tests to ensure that the function fails if the requisite number of instances aren't available after the given retry period has elapsed. The function now also has optional arguments that allow you to specify what those periods are. Let me know if this handles your thoughts?

File: test/aws-wrapper.bats (hunk 7 above)
[the remark this answers is not in the document]
Comment (brikis98) [reply test/aws-wrapper.bats]: Thanks! I thought we cleaned these all up already, but I guess not.

File: Dockerfile.ubuntu16.04.bats [hunk 9]
```diff
@@ -3,7 +3,11 @@ MAINTAINER Gruntwork <info@gruntwork.io>
 
 # Install basic dependencies
 RUN DEBIAN_FRONTEND=noninteractive apt-get update && \
-    apt-get install -y vim python-pip jq sudo curl
+    apt-get install -y vim python-pip jq sudo curl libffi-dev python3-dev && \
+    update-alternatives --install /usr/bin/python python /usr/bin/python3 1 && \
```
Comment (brikis98) [inline Dockerfile.ubuntu16.04.bats:7]: Sanity check... Is the `1` after `/usr/bin/python3` intentional?

File: Dockerfile.ubuntu16.04.bats [hunk 10]
```diff
@@ -25,3 +29,7 @@ RUN apt-get install -y net-tools iptables
 
 # Copy mock AWS CLI into the PATH
 COPY ./.circleci/aws-local.sh /usr/local/bin/aws
+
+# These have been added to resolve some encoding issues with the tests
```
Comment (brikis98) [inline Dockerfile.ubuntu16.04.bats:33]: NIT: is there some GitHub issue or SO thread to explain the context here?

**Review by brikis98 — COMMENTED** [review]
Very close now. Just a couple final questions.

File: Dockerfile.ubuntu16.04.bats (hunk 10 above)
Comment (yardbirdsax) [reply Dockerfile.ubuntu16.04.bats:33]: Not a specific one, more like a combination of looking at some of the errors generated after upgrading to Python 3 and then some google-fu.

Here's an example of an error that happens after switching to Python 3 (but before these env variables are added):
```
bats_ubuntu1804_1  | not ok 57 aws_get_instance_tags empty
...
bats_ubuntu1804_1  | #     return codecs.ascii_decode(input, self.errors)[0]
bats_ubuntu1804_1  | # UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 13950: ordinal not in range(128)
```

That lead to finding threads like [this one](https://stackoverflow.com/questions/51026315/how-to-solve-unicodedecodeerror-in-python-3-6/51027262#51027262) which is where the environment variables have come from. I'll add some more details here and push that shortly.

File: Dockerfile.ubuntu16.04.bats (hunk 9 above)
Comment (brikis98) [reply Dockerfile.ubuntu16.04.bats:7]: Roger

File: Dockerfile.ubuntu16.04.bats (hunk 10 above)
Comment (brikis98) [reply Dockerfile.ubuntu16.04.bats]: Perfect, thanks for adding that!

**Review by brikis98 — APPROVED** [review]
Code changes LGTM, thanks! I'll kick off tests now.

**Comment by brikis98** [comment]
Tests passed! Merging now.

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/bash-commons/releases/tag/v0.1.4

**Pull request merged**